Ndev 2477 add send neon load test scenario

romanova-natasha opened the pull request.

romanova-natasha committed:
add local run section to readme

add local run section to readme

romanova-natasha committed:
add fundAccount function with an opportunity to choose a source: faucet or bank account or any other custom func

add fundAccount function with an opportunity to choose a source: faucet or bank account or any other custom func

romanova-natasha committed:
add sender private key to client

add sender private key to client

romanova-natasha committed:
reorganize metrics logic: mv all eth methods metrics to the plugin

reorganize metrics logic: mv all eth methods metrics to the plugin

romanova-natasha committed:
fix import

fix import

romanova-natasha committed:
sign tx with sender account

sign tx with sender account

romanova-natasha committed:
add send neon scenario options for load run

add send neon options for load run

romanova-natasha committed:
add scenario options description to readme

add scenario options description to readme

kristinaNikolaevaa commented on loadtesting/k6/tests/sendNeon.test.js:
```diff
@@ -1,121 +1,44 @@
-    }
+    const accountSender = createAccount();
+    const client = ethClient(accountSender.private_key);
+    fundAccountFromFaucet(client, accountSender.address, initialAccountBalance);
```
maybe it's better to prepare accounts first? airdrop neons to sender and do only transfer in this function? Looks like we test faucet here under load

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on loadtesting/k6/tests/utils/ethClient.js:
```diff
@@ -38,13 +35,19 @@ export function sendTokens(client, from, to, value, gas, gasPrice, nonce) {
-    if (transaction["gas"] == 0) {
-        transaction["gas"] = client.estimateGas(transaction);
+    if (gas == null) {
+        transaction["gas"] = client.estimateGas({
```
why don't you use transaction object here as it was?

romanova-natasha replied:
you're right, it was just old debug code
fixed

romanova-natasha committed:
add prepare accounts step before run, add shared context for VUs, setup k6 envs

add prepare accounts step before run, add shared context for VUs, setup k6 envs

romanova-natasha committed:
rm dbg

rm dbg

kristinaNikolaevaa commented on clickfile.py:
```diff
@@ -1253,10 +1255,44 @@ def build(tag):
+                       "sender_key": str(account_sender.key.hex())[2:], 
+                       "receiver_address": str(account_receiver.address)}
+
+    with open('./loadtesting/k6/data/accounts.json', 'w', encoding='utf-8') as f:
```
lets move it from clickfile.py to another script  to have here only run command 
because except accounts we need deployed contracts and maybe some another prepared data in future
It is better not to overload clickfile

romanova-natasha replied:
agree

romanova-natasha replied:
moved to utils/k6_helpers.py

kristinaNikolaevaa commented on clickfile.py:
```diff
@@ -1266,16 +1302,38 @@ def run(network, script):
+    if os.environ.get('TRACER_URL') is None:
+        os.environ["TRACER_URL"] = config[network]['tracer_url']
+
+    if os.environ.get('SPL_NEON_MINT') is None:
```
why do you need a possibility to replace all env variables from  local environment instead of envs.json ?  we have there all networks,  it is more transparent to use urls and addresses as it used for another tests

romanova-natasha replied:
I get all params from envs.json and set up everything as envs for k6, it is how k6 gets needed info.

kristinaNikolaevaa replied:
Yes, i see
but why do you need conditions like this
if os.environ.get('SPL_NEON_MINT') is None:

romanova-natasha replied:
all unused evns are removed

kristinaNikolaevaa commented on loadtesting/k6/data/accounts.json:
```diff
@@ -0,0 +1,12 @@
+{
+    "0": {
+        "sender_address": "0x64f387a5b5fAe44C1fC00a5FB5328E09a94166BE",
```
I think this file is committed accidentally

romanova-natasha replied:
fixed

romanova-natasha replied:
k6 data folder is also added to gitignor

kristinaNikolaevaa commented on loadtesting/k6/tests/utils/accountManager.js:
```diff
@@ -13,7 +13,24 @@ export function createAccount() {
 }
 
-export function fundAccountWithNeon(address, amount) {
+export function fundAccountFromFaucet(client, address, amount) {
```
so if you moved accounts preparing to clickfile do you still need these 3 functions here?

romanova-natasha replied:
removed

romanova-natasha committed:
mv preparations stages to k6_helpers.py

mv preparations stages to k6_helpers.py

romanova-natasha committed:
rm unused envs

rm unused envs

romanova-natasha committed:
add k6 data folder to gitignor

add k6 data folder to gitignor

romanova-natasha committed:
rm account manager

rm account manager

kristinaNikolaevaa commented on utils/k6_helpers.py:
```diff
@@ -0,0 +1,66 @@
+    if os.environ.get('FAUCET_URL') is None:
+        os.environ["FAUCET_URL"] = config[network]['faucet_url']
+
+    if os.environ.get('NETWORK_ID') is None:
```
I still suggest to use values from config without conditions

os.environ["PROXY_URL"] = config[network]['proxy_url']
os.environ["SOLANA_URL"] = config[network]['solana_url']
os.environ["FAUCET_URL"] = config[network]['faucet_url']
...

Because it is a potential problem for local run if there are some old value

romanova-natasha replied:
fixed
I get proxy url from envs json in k6 script

kristinaNikolaevaa commented on utils/k6_helpers.py:
```diff
@@ -0,0 +1,66 @@
+        os.environ["NETWORK_ID"] = str(config[network]['network_ids']['neon'])
+    
+    # This parameter is mandatory for k6 scenario setup
+    if os.environ.get('K6_USERS_NUMBER') is None:
```
And here too
if you run command with --users=200 you want to use this value, not old value from K6_USERS_NUMBER which was set earlier

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on utils/k6_helpers.py:
```diff
@@ -0,0 +1,66 @@
+        if initial_balance is not None:
+            os.environ["K6_INITIAL_BALANCE"] = initial_balance
+        else:
+            print("Users initial balance is not set. Please set K6_INITIAL_BALANCE env or pass it as an argument to the k6 run command.")
```
you have required=True for balance option for the command
This part of code impossible to reach

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on utils/k6_helpers.py:
```diff
@@ -0,0 +1,66 @@
+
+def k6_prepare_accounts(network, users, balance, bank_account):
+    print("Setting k6 envs...")
+    config = k6_set_envs(network, "envs.json", users, balance, bank_account)
```
you can use class NetworkManager which was created for reading network params from this file

romanova-natasha replied:
done

romanova-natasha committed:
change env settings, rm unused blocks of code, get params for k6 script from envs.json

change env settings, rm unused blocks of code, get params for k6 script from envs.json

romanova-natasha committed:
get proxy_url from envs.json directly

get proxy_url from envs.json directly

romanova-natasha committed:
Merge branch 'develop' into NDEV-2477-add-send-neon-load-test-scenario
